src/multiplicity_vectors.lean

```diff
@@ -0,0 +1,645 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson.
+-/
+
+import data.finsupp
+import data.pnat.factors
+import order.conditionally_complete_lattice
+import order.lattice
+import order.order_iso
+import tactic.pi_instances
+
+/-!
+# Factor Finsupps
+
+A new version of prime factorisation and some facts about lattice isomorphisms
+
+## Notations
+
+## Implementation Notes
+factor_finsupp is similar to pnat.factor_multiset, but a finsupp of exponents
+Most of the sorries will be solved by proving some facts about lattice isomorphisms.
+
+## References
+
+-/
+
+set_option old_structure_cmd true
+
+open_locale classical
+open_locale big_operators
+
+
+noncomputable theory
+
+section basics
+
+def pnat.coprime (m n : ℕ+) : Prop := m.gcd n = 1
+
+def mult (n : ℕ+) (p : nat.primes) : ℕ := (pnat.factor_multiset n).count p
+
+def prime_finsupp : Type := nat.primes →₀ ℕ
+
+@[instance]
+instance prime_finsupp.has_coe_to_fun : has_coe_to_fun prime_finsupp := { ..finsupp.has_coe_to_fun}
+
+def prime_finsupp.to_pnat_finsupp : prime_finsupp → ℕ+ →₀ ℕ := finsupp.map_domain coe
+
+@[instance]
+instance prime_finsupp.coe_pnat : has_coe prime_finsupp (ℕ+ →₀ ℕ) :=
+⟨prime_finsupp.to_pnat_finsupp⟩
+
+lemma prime_finsupp.coe_pnat_to_multiset (x : prime_finsupp) :
+  prime_multiset.to_pnat_multiset x.to_multiset = (x : ℕ+ →₀ ℕ).to_multiset :=
+finsupp.to_multiset_map x coe
+
+def prime_finsupp.prod_apply_pow {α : Type} [comm_monoid α] (x : prime_finsupp) (f : ℕ+ → α) : α :=
+x.prod (λ p : nat.primes, λ k : ℕ, f (p ^ k))
+
+def prime_finsupp.prod_pow (x : prime_finsupp) := x.prod_apply_pow id
+
+lemma prime_finsupp.prod_pow_eq (x : prime_finsupp) :
+  x.prod_pow= x.prod (λ p : nat.primes, λ k : ℕ, (p ^ k)) := rfl
+
+lemma prime_finsupp.prod_pow_eq_coe_prod_pow (x : prime_finsupp) :
+  x.prod_pow= (x : ℕ+ →₀ ℕ).prod pow :=
+begin
+  change x.prod_pow = (finsupp.map_domain coe x).prod pow,
+  rw [finsupp.prod_map_domain_index, prime_finsupp.prod_pow_eq], simp,
+  intros, rw pow_add,
+end
+
+lemma prime_finsupp.prod_pow_eq_prod_to_multiset (x : prime_finsupp) :
+  x.prod_pow= prime_multiset.prod x.to_multiset :=
+begin
+  rw [prime_multiset.prod, prime_finsupp.prod_pow_eq_coe_prod_pow,
+    ← finsupp.prod_to_multiset, ← prime_finsupp.coe_pnat_to_multiset], refl
+end
+
+def factor_finsupp (n : ℕ+) : prime_finsupp := n.factor_multiset.to_finsupp
+
+@[simp]
+lemma inf_zero_iff {m n : ℕ} : m ⊓ n = 0 ↔ m = 0 ∨ n = 0 :=
+begin
+  split, swap, intro h, cases h; rw h; simp,
+  unfold has_inf.inf, unfold semilattice_inf.inf, unfold lattice.inf, unfold min,
+  intro hmin, by_cases m = 0, left, apply h,
+  rw if_neg _ at hmin, right, apply hmin, contrapose hmin,
+  simp only [not_lt, not_le] at hmin, rw if_pos hmin, apply h
+end
+
+lemma nat.zero_eq_bot : (0 : ℕ) = ⊥ := rfl
+
+variable {α : Type}
+
+instance finsupp.has_inf :
+  has_inf (α →₀ ℕ) :=
```

Comment:
I don't think that we want this to be the global `has_inf` instance for finsupps. I guess you can replace `nat` with a certain type of lattice `X` and deduce some other lattice structure for `\a →\0 X`.
I haven't checked how much of this is already done in the library.